File: c++/src/node.cpp
#include <string>
#include <iostream>
#include <cstdlib>
#include <vector>
#include <node.h>

void Node::insert(Node * node) {
  if (node->key < this->key) {
    insert_left(node);
  } else {
    insert_right(node);
  }
}

void Node::unlink() {
  left = right = parent = nullptr;
}

bool Node::is_unlinked() {
  return left == nullptr && right == nullptr && parent == nullptr;
}

void Node::insert_left(Node * node) {
  if (this->left == nullptr) {
    node->parent = this;
    this->left = node;
  } else {
    left->insert(node);
  }
}

void Node::insert_right(Node * node) {
  if (this->right == nullptr) {
    node->parent = this;
    this->right = node;
  } else {
    right->insert(node);
  }
}

Node * Node::search(int key) {
  if (this->key == key) return this;

  if (key < this->key) {
    if (left != nullptr) {
      return left->search(key);
    } else {
      return right->search(key);
    }
  }
  return nullptr;
}


bool Node::get_is_bst(int minimum, bool result) {
  if (this->left != nullptr) { result = this->left->get_is_bst(minimum, result); }
  if (minimum >= this->key) { return false; }
  minimum = this->key;
  if (this->right != nullptr) { result = this->right->get_is_bst(minimum, result); }
  return result;
}

bool Node::is_bst() {
  return get_is_bst(-10000, true);
}

bool Node::is_present(int key) {
  if (this->key == key) return true;

  if (this->left != nullptr && key < this->left->key) {
    return left->is_present(key);
  } else if (this->right != nullptr) {
    return right->is_present(key);
  }
  return false;
}

Node * Node::get_successor(Node * node, Node * parent, Node * successor) {
  if (parent->left != nullptr && parent->left == this) { successor = parent; }

  if (this == node) {
    if (this->right != nullptr) {
      return this->right->minimum();
    } else {
      return successor;
    }
  }

  if (node->key < this->key) {
    if (this->left != nullptr) {
      return this->left->get_successor(node, this, successor);
    }
  } else {
    if (this->right != nullptr) {
      return this->right->get_successor(node, this, successor);
    }
  }
}

Node * Node::successor(Node * node) {
  return get_successor(node, this, node);
}

Node * Node::get_predecessor(Node * node, Node * parent, Node * predecessor) {
  if (parent->right != nullptr && parent->right == this) { predecessor = parent; }

  if (this == node) {
    if (this->left != nullptr) {
      return this->left->maximum();
    } else {
      return predecessor;
    }
  }

  if (node->key < this->key) {
    if (this->left != nullptr) {
      return this->left->get_predecessor(node, this, predecessor);
    }
  } else {
    if (this->right != nullptr) {
      return this->right->get_predecessor(node, this, predecessor);
    }
  }
}

Node * Node::predecessor(Node * node) {
  return get_predecessor(node, this, node);
}

void Node::post_order_traverse(std::function<void (void)> callback) {
  if (left != nullptr) { left->post_order_traverse(callback); }
  if (right != nullptr) { right->post_order_traverse(callback); }
  callback();
}

int Node::get_height(int height, int max) {
  int current = ++height;
  if (left != nullptr) { max = left->get_height(current, max); }
  if (right != nullptr) { max = right->get_height(current, max); }
  current--;
  return current > max ? current : max;
}

int Node::height(void) {
  return get_height(0, 0);
}

void Node::in_order_traverse(std::function<void (void)> callback) {
  if (left != nullptr) { left->in_order_traverse(callback); }
  callback();
  if (right != nullptr) { right->in_order_traverse(callback); }
}

std::vector<int> Node::list_keys() {
  std::vector<int> keys;
  collect(keys);
  return keys;
}

// TODO: get rid of get_left and get_right
void Node::get_left(std::vector<int> & keys) {
  if (left == nullptr) {
    return;
  } else {
    left->collect(keys);
  }
}

void Node::get_right(std::vector<int> & keys) {
  if (right == nullptr) {
    return;
  } else {
    right->collect(keys);
  }
}

void Node::collect(std::vector<int> & keys) {
  get_left(keys);
  keys.push_back(key);
  get_right(keys);
}

int Node::size(void) {
  int size = 0;
  post_order_traverse([&]{ size++; });
  return size;
}

Node * Node::minimum(void) {
  if (left == nullptr) return this;
  return this->left->minimum();
}

Node * Node::maximum(void) {
  if (right == nullptr) return this;
  return this->right->maximum();
}

std::string
get_node(void) {

  return std::string("node");
}

File: python/lib/node.py
class Node(object):

    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        # p is pointer to parent node. The goal of this class is
        # to implement as much as possible without an explicit reference
        # to the parent node. However, some algorithms require the
        # parent node (e.g., clr), so it's defined as a member here to
        # be used for implementations of those algorithms. Hopefully,
        # better algorithms can be derived and the parent pointer
        # elmininated from this class.
        # UPDATE: changing, s/p/parent/ for better readability.
        self.parent = None

    def collect(self, collector):
        if self.left is None:
            pass
        else:
            self.left.collect(collector)

        collector.append(self.key)

        if self.right is None:
            pass
        else:
            self.right.collect(collector)

    def insert(self, node):
        if node.key < self.key:
            if self.left is None:
                node.parent = self
                self.left = node
            else:
                self.left.insert(node)
        else:
            if self.right is None:
                node.parent = self
                self.right = node
            else:
                self.right.insert(node)

    def find(self, key):
        if self.key == key:
            return self

        if key < self.key:
            if self.left is not None:
                return self.left.find(key)
        else:
            if self.right is not None:
                return self.right.find(key)

    def find_with_parent(self, key, parent=None):
        print "in find_with_parent, key: %d" % key
        if self.key == key:
            return self, parent

        if key < self.key:
            if self.left is not None:
                return self.left.find_with_parent(key, self)
        else:
            if self.right is not None:
                return self.right.find_with_parent(key, self)


    def is_present(self, key):
        if self.key == key:
            return True

        if key < self.key:
            if self.left is not None:
                return self.left.is_present(key)
        else:
            if self.right is not None:
                return self.right.is_present(key)

    def compute_size(self):
        def get_size(size):
            size += 1
            if self.left is not None:
                gs = self.left.compute_size()
                size = gs(size)
            if self.right is not None:
                gs = self.right.compute_size()
                size = gs(size)
            return size
        return get_size

    def size(self):
        cs = self.compute_size()
        return cs(0)

    def get_successor(self, node, parent, successor):
        if parent.left is not None:
            if parent.left.key == self.key:
                successor = parent

        if node.key == self.key:
            if node.right is not None:
                return node.right.minimum()
            else:
                return successor

        if node.key < self.key:
            if self.left is not None:
                return self.left.get_successor(node, self, successor)
        else:
            if self.right is not None:
                return self.right.get_successor(node, self, successor)


    def successor(self, node):
        return self.get_successor(node, self, node)

    def get_predecessor(self, node, parent, predecessor):
        if parent.right is not None:
            if parent.right.key == self.key:
                predecessor = parent

        if node.key == self.key:
            if node.left is not None:
                return node.left.maximum()
            else:
                return predecessor

        if node.key < self.key:
            if self.left is not None:
                return self.left.get_predecessor(node, self, predecessor)
        else:
            if self.right is not None:
                return self.right.get_predecessor(node, self, predecessor)

    def predecessor(self, node):
        return self.get_predecessor(node, self, node)

    # TODO: refactor this to work with a generic in_order_traverse
    def is_bst(self, minimum=-1000, result=True):
        if self.left is not None:
            result = self.left.is_bst(minimum, result)

        if minimum >= self.key:
            result = False
            return result

        minimum = self.key

        if self.right is not None:
            result = self.right.is_bst(minimum, result)

        return result

    # algorithm originally taken from CLR p. 253
    # parent pointers are part of CLR's definition of
    # node in BST.
    def clr_delete(T, node):
        z, p = T.root.find_with_parent(node)

        if z.left is None or z.right is None:
            y = z
        else:
            y = z.successor(z)

        if y.left is not None:
            x = y.left
        else:
            x = y.right

        if x is not None:
            x.p = y.p # p is link to parent node

        if y.p is None:
            T.root = x
        else:
            if y == y.p.left:
                y.p.left = x
            else:
                y.p.right = x

        if y != z:
            z.key = y.key

        return y

    def maximum(self):
        if self.right is None:
            return self
        else:
            return self.right.maximum()

    def minimum(self):
        if self.left is None:
            return self
        else:
            return self.left.minimum()
